dotnetcore3.1/cookiecutter-aws-sam-quick-start-web-dotnet/{{cookiecutter.project_name}}/src/{{cookiecutter.project_name}}/Controllers/BooksController.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+using Microsoft.AspNetCore.Mvc;
+
+using Amazon.DynamoDBv2;
+using Amazon.DynamoDBv2.DataModel;
+using Amazon.Lambda.Core;
+
+using {{cookiecutter.project_name}}.Entities;
+using Amazon.DynamoDBv2.DocumentModel;
+
+namespace {{cookiecutter.project_name}}.Controllers
+{
+    [Route("api/[controller]")]
+    public class BooksController : ControllerBase
```

Comment:
AFAIU, you created a single handler for your function, and created each route inside that function within a possible middleware. So your middleware contains route configuration.

I am not sure if this is a best practice for dotnet, but I usually see the other-way around for other languages.

Usually, individual route configuration created inside `template.yaml` and each route will call a different function for its execution. For instance `/` route will call a function (lets say `GetAllBooks`) and that function will return all books from DDB table. So the middleware (which sets the routes and handlers) is actually ApiGW itself.

Can we change this init template into other format where each route will have its own function?

Reply:
@mndeveci 

This is the best practice for .NET Lambda Web project. Doing it like this keeps it consistence with [.NET Core CLI](https://docs.aws.amazon.com/lambda/latest/dg/csharp-package-cli.html), the lambda team provides this CLI for .NET developers to create AWS Lambda projects with multiples templates including the WebApi. You can teste it by running 
```bash
dotnet new serverless.AspNetCoreWebAPI --name lab3
```
This should create a new project called `lab3` with the same structure, structuring the .NET Lambda WebAPI like this gives much better performance and reduce the cold start drastically if compared with the way you’re suggesting.

Reply:
I see, if that is what is been provided by dotnetcli then it is fine. My other concern was around TPS limitation. Because if we go in this direction, your total TPS for all endpoints will be 1k by default, and they need to share that. But if you would have different lambda definitions, each handling different endpoit, then each endpoint will have 1k TPS.